README.md
## WooCommerce Recently Viewed Products

WooCommerce Recently Viewed Products adds a section in your WooCommerce shop to displays a visitors recently viewed Products; better than other plugins like this because its light, more modern and relies mostly on native WooCommerce functionality.

## How it works

![Example 1](screenshots/example-1.png?raw=true "Example 1")

When on product pages, recently viewed products are visible.

## Features

All of the templates in the templates/ folder are overridable by including the file(s) in your theme folder.

The plugin stores the Product ID's in the wp options table for guests, and stores the Product ID's in wp usermeta table for logged in users.

## Installation

1. Upload the plugin to the **/wp-content/plugins/** directory
2. Activate the plugin through the 'Plugins' menu in WordPress

## Requirements

PHP 5.4+

Wordpress 4+

WooCommerce 2.5+

## License

[GNU General Public License v3.0](http://www.gnu.org/licenses/gpl-3.0.html)

changelog.txt
*** WooCommerce Recently Viewed ***

2015.04.23 - version 1.0.0
 * Initial Release.


templates/content-recently-viewed-products.php
<?php

/**
 * Recently viewed products template.
 *
 * @version 	1.0.0
 */

if ( ! defined( 'ABSPATH' ) ) {
	exit; // Exit if accessed directly
}

global $woocommerce_recently_viewed_products;

?>

<?php if( have_posts() ) : ?>

	<?php
		/**
		 * woocommerce_before_shop_loop hook
		 *
		 * @hooked woocommerce_result_count - 20
		 * @hooked woocommerce_catalog_ordering - 30
		 */
		do_action( 'woocommerce_before_recently_viewed_products_loop' );
	?>
	
	<h1 class="<?php echo implode(' ', apply_filters('woocommerce_before_recently_viewed_products_title_classes', array('title'))); ?>"><?php echo apply_filters('woocommerce_before_recently_viewed_products_title', __('Recently viewed by you', 'woocommerce-recently-viewed-products')); ?></h1>
	
	<?php woocommerce_product_loop_start(); ?>
	
	<?php while( have_posts() ) : the_post(); ?>
	
		<?php wc_get_template(apply_filters('woocommerce_recently_viewed_product_template', 'content-product.php')); ?>
	
	<?php endwhile; ?>
	
	<?php woocommerce_product_loop_end(); ?>

	<?php
		/**
		 * woocommerce_after_shop_loop hook
		 *
		 * @hooked woocommerce_pagination - 10
		 */
		do_action( 'woocommerce_after_recently_viewed_products_loop' );
	?>

<?php elseif ( ! woocommerce_product_subcategories( array( 'before' => woocommerce_product_loop_start( false ), 'after' => woocommerce_product_loop_end( false ) ) ) && apply_filters('woocommerce_after_recently_viewed_products_display_no_products_found', true) ) : ?>

	<?php wc_get_template( 'loop/no-products-found.php' ); ?>

<?php endif; wp_reset_query(); ?>

woocommerce-recently-viewed-products.php
<?php
/*
* Plugin Name: WooCommerce Recently Viewed Products
* Description: Adds a section in your shop that displays a visitors recently viewed products.
* Version: 1.0.0
* Author: Creative Little Dots
* Author URI: http://creativelittledots.co.uk
*
* Text Domain: woocommerce-recently-viewed-products
* Domain Path: /languages/
*
* Requires at least: 3.8
* Tested up to: 4.1.1
*
* Copyright: © 2009-2015 Creative Little Dots
* License: GNU General Public License v3.0
* License URI: http://www.gnu.org/licenses/gpl-3.0.html
*/

// Exit if accessed directly
if ( ! defined( 'ABSPATH' ) ) {
	exit;
}

if( ! function_exists('is_woocommerce_active') ) {
	return;
}

// Check if WooCommerce is active
if ( ! is_woocommerce_active() ) {
	
	return;
	
}


class WC_Recently_Viewed {
	
	public $version 	= '1.0.0';
	
	public function __construct() {
		
		add_action( 'init', array( $this, 'init' ) );
		add_action( 'admin_init', array( $this, 'activate' ) );
		register_deactivation_hook( __FILE__, array( $this, 'deactivate' ) );
		
		// Add Recently Viewed Products to Shop Page
		add_action( 'woocommerce_after_main_content', array($this, 'woocommerce_recently_viewed_products') );
		
		// Tracking Product View
		add_action( 'wp', array($this, 'woocommerce_recently_viewed_products_track_view') );
		
		// Before Recently View Loop
		add_action('woocommerce_before_recently_viewed_products_loop', 'woocommerce_result_count', 20);
		add_action('woocommerce_before_recently_viewed_products_loop', 'woocommerce_catalog_ordering', 30);
		
		// After Recently Viewed Loop
		add_action('woocommerce_after_recently_viewed_products_loop', 'woocommerce_pagination');
		
	}
	
	public function init() {
		load_plugin_textdomain( 'woocommerce-recently-viewed', false, dirname( plugin_basename( __FILE__ ) ) . '/languages/' );
	}
	
	public function activate() {
		
		global $wpdb;

		$version = get_option( 'woocommerce_recently_viewed_products_version', false );
		
		if ( $version == false ) {
			
			add_option( 'woocommerce_recently_viewed_products_version', $this->version );

			// Update from previous versions

			// delete old option
			delete_option( 'woocommerce_composite_products_extension_active' );
				
		} elseif ( version_compare( $version, $this->version, '<' ) ) {

			update_option( 'woocommerce_recently_viewed_products_version', $this->version );
		}

	}
	
	/**
	 * Deactivate extension.
	 * @return void
	 */
	public function deactivate() {

		delete_option( 'woocommerce_recently_viewed_products_version' );
		
	}
	
	public function plugin_url() {
		return plugins_url( basename( plugin_dir_path(__FILE__) ), basename( __FILE__ ) );
	}

	public function plugin_path() {
		return untrailingslashit( plugin_dir_path( __FILE__ ) );
	}

	public function plugins_loaded() {
		
		global $woocomerce;	
		
	}
	
	public function woocommerce_recently_viewed_products() {
		
		global $woocommerce_recently_viewed_products, $current_user;
		
		$recently_viewed_products = array(0);
		
		if(is_user_logged_in()) {
			
			if($current_user->woocommerce_recently_viewed_products && is_array($current_user->woocommerce_recently_viewed_products)) {
				
				$recently_viewed_products = $current_user->woocommerce_recently_viewed_products;
				
			}
			
			elseif($recently_viewed_products_by_ip_address = $this->woocommerce_recently_viewed_products_get_by_ip_address($_SERVER['REMOTE_ADDR'])) {
				
				$recently_viewed_products = $recently_viewed_products_by_ip_address;
				
			}
			
		}
		
		elseif($recently_viewed_products_by_ip_address = $this->woocommerce_recently_viewed_products_get_by_ip_address($_SERVER['REMOTE_ADDR'])) {
			
			$recently_viewed_products = $recently_viewed_products_by_ip_address;
			
		}
		
		query_posts(apply_filters('woocommerce_recently_viewed_products_args', array(
		    'post_type' => 'product',
		    'post__in'	=> $recently_viewed_products,
		    'orderby' => 'post__in',
		)));
		
		$args = array();
	
		wc_get_template('content-recently-viewed-products.php', $args, false, $woocommerce_recently_viewed_products->plugin_path() . '/templates/');
		
	}
	
	public function woocommerce_recently_viewed_products_get_by_ip_address($ip_address) {
		
		$recently_viewed_products = get_option('woocommerce_recently_viewed_products');
		
		if(is_array($recently_viewed_products) && $recently_viewed_products) {
			
			if(isset($recently_viewed_products[$ip_address]) && $recently_view_products_by_user = $recently_viewed_products[$ip_address]) {
				
				return $recently_view_products_by_user;
				
			}
			
		}
		
		else {
			
			return false;
			
		}
		
	}
	
	public function woocommerce_recently_viewed_products_track_view() {
		
		if(is_singular('product')) {
			
			global $post;
			
			$recently_viewed_products = array();
			
			$product = new WC_Product($post->ID);
			
			if($product->is_visible()) {
				
				if(is_user_logged_in()) {
					
					global $current_user;
					
					$recently_viewed_products = $current_user->woocommerce_recently_viewed_products;
					
				}
				
				else {
					
					$recently_viewed_products = $this->woocommerce_recently_viewed_products_get_by_ip_address($_SERVER['REMOTE_ADDR']);
					
				}
				
				$recently_viewed_products = is_array($recently_viewed_products) && $recently_viewed_products ? $recently_viewed_products : array();
				
				array_unshift($recently_viewed_products, $product->id);
				
				$recently_viewed_products = array_unique($recently_viewed_products);
				
				if(is_user_logged_in()) {
					
					update_user_meta($current_user->ID, 'woocommerce_recently_viewed_products', $recently_viewed_products);
					
				}
				
				else {
					
					update_option('woocommerce_recently_viewed_products', array_merge(get_option('woocommerce_recently_viewed_products') ? get_option('woocommerce_recently_viewed_products') : array(), array($_SERVER['REMOTE_ADDR'] => $recently_viewed_products)));
					
				}
				
			}
			
		}
		
	}
	
}

$GLOBALS[ 'woocommerce_recently_viewed_products' ] = new WC_Recently_Viewed();